src/NHibernate/Proxy/DynamicProxy/ProxyCacheEntry.cs

```diff
@@ -24,6 +24,7 @@ public ProxyCacheEntry(System.Type baseType, System.Type[] interfaces)
 			}
 			BaseType = baseType;
 			Interfaces = interfaces ?? new System.Type[0];
+			_uniqueInterfaces = new HashSet<System.Type>(Interfaces);
```

Comment:
I think, what we need is presort `Interfaces` (eg. `interfaces.Distinct().OrderBy(x=>x.FullName).ToArray()`).